### ISSUE TITLE
two instances to different servers cannot keep connection at the same time - second connection will close first connection

### ISSUE THREAD

**Issue by Killerjoe** [domschl/python-fhem#26]
Hi,

first of all, thank you for the python module for fhem.
I like to use it, but have issues with my setup.

I have two separate fhem installations (fhem1 and fhem2) and wanted to query all of my pca301 sockets out of both.

I were creating two connections to fhem1 and fhem2:

```
fh1 = fhem.Fhem("fhem1", port = "8084", protocol = "https", use_ssl = True, username = "test1", password = 'XYZ', csrf=True)
fh2 = fhem.Fhem("fhem2", port = "8084", protocol = "https", use_ssl = True, username = "test1", password = 'XYZ', csrf=True)

fh1.connected()
fh2.connected()
>>> fh1.connected()
False
>>> fh1.connect()
>>> fh1.connected()
True
>>> fh1.get_readings(name="PCA301_WZ_TV_AVR1")
{'PCA301_WZ_TV_AVR1': {'IODev': {'Value': 'Pwrjee', 'Time': datetime.datetime(2023, 8, 8, 22, 49, 13)}, 'consumption': {'Value': 2.2, 'Time': datetime.datetime(2023, 8, 9, 18, 22, 58)}, 'consumptionTotal': {'Value': 1385.98999999927, 'Time': datetime.datetime(2023, 8, 9, 18, 22, 58)}, 'power': {'Value': 16.7, 'Time': datetime.datetime(2023, 8, 9, 18, 28, 12)}, 'state': {'Value': 'on', 'Time': datetime.datetime(2023, 8, 9, 0, 45, 17)}, 'waitforACK': {'Value': 'ACK:on', 'Time': datetime.datetime(2023, 8, 5, 22, 1, 54)}}}
>>>
>>> fh2.connected()
False
>>> fh2.connect()
>>> fh2.connected()
True
>>> fh2.get_readings(name="PCA301_Buero_MiniPC")
{'PCA301_Buero_MiniPC': {'IODev': {'Value': 'Pwrjee', 'Time': datetime.datetime(2023, 8, 8, 22, 5, 58)}, 'consumption': {'Value': 1.57, 'Time': datetime.datetime(2023, 8, 9, 18, 17, 50)}, 'consumptionTotal': {'Value': 5143.26000000688, 'Time': datetime.datetime(2023, 8, 9, 18, 17, 50)}, 'power': {'Value': 15.8, 'Time': datetime.datetime(2023, 8, 9, 18, 30, 21)}, 'state': {'Value': 'on', 'Time': datetime.datetime(2023, 3, 20, 15, 32, 54)}}}
>>> fh1.get_readings(name="PCA301_WZ_TV_AVR1")
Failed to send msg, len=32, HTTP Error 401: Authorization Required
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/scripts/power/powerenv/lib/python3.10/site-packages/fhem/__init__.py", line 597, in get_readings
    return self._response_filter(response, arg, 'Readings', value_only=value_only, time_only=time_only)
  File "/scripts/power/powerenv/lib/python3.10/site-packages/fhem/__init__.py", line 460, in _response_filter
    for r in response if 'totalResultsReturned' not in response else response['Results']:
TypeError: argument of type 'NoneType' is not iterable
>>> fh1.connected()
False
>>> 
```

As you can see, both can connect and query a socket from fhem1 (PCA301_WZ_TV_AVR1) or fhem2 (PCA301_Buero_MiniPC) successful.
But after I have connected to the second server fhem2, I will lose my connection to the first server fhem1 and cannot query again the socket PCA301_WZ_TV_AVR1 and get the error:
`for r in response if 'totalResultsReturned' not in response else response['Results']:
TypeError: argument of type 'NoneType' is not iterable`

All this happend within 1 minute, so it is no timeout or something.
I also have tried it reverse, so started first with fhem2 and then fhem1 and then will lose connection to fhem2, so the first server, I have made the connection.

Can you please check why this is happening and fix the issue?

Regards,
Jochen

**Comment by domschl** [comment]
I've looked again (after quite some time) through the code, and there doesn't seem to be anything that points to a global state (which could be corrupted by multiple instances). Basically, the two different connections [in theory!] have no means of knowing of each other's existence.

So I suspect that something else is going wrong. Are you sure that the connection remains stable if you just use a single server repeatedly?

**Comment by Killerjoe** [comment]
I also had a look into your code and I don´t understand why it happens.
One session alone is stable and will not be interrupted.

I just did a for loop in range 20 and was getting data from one instance.
After each iteration, I was waiting 5 seconds and the loop starts again. I had no connection issues.
```
for _ in range(20):
    fh1.get_readings(name="PCA301.*")
    time.sleep(5)
```

But again, after I have created the second connection with:
`fh2 = fhem.Fhem("fhem2", port = "8084", protocol = "https", use_ssl = True, username = "test1", password = 'XYZ', csrf=True)`
I was unable to get an answer from fh1. I was getting again the error:
`Failed to send msg, len=23, HTTP Error 401: Authorization Required`

Here you can see both information from the objects at that time.
From the memory objects it looks different, so not really understandable what happens.

```
fh1.__dict__
{'log': <Logger Fhem (WARNING)>, 'server': 'fhem1', 'port': '8084', 'ssl': True, 'csrf': True, 'csrftoken': 'csrf_496429725763431', 'username': 'test1', 'password': 'XYZ', 'loglevel': 1, 'connection': True, 'cafile': '', 'nolog': False, 'bsock': None, 'sock': None, 'https_handler': <urllib.request.HTTPSHandler object at 0x7f466af01f60>, 'protocol': 'https', 'baseurlauth': 'https://fhem1:8084/', 'baseurltoken': 'https://fhem1:8084/fhem', 'baseurl': 'https://fhem1:8084/fhem?XHR=1&cmd=', 'opener': <urllib.request.OpenerDirector object at 0x7f466af01f00>, 'password_mgr': <urllib.request.HTTPPasswordMgrWithDefaultRealm object at 0x7f466af01ff0>, 'auth_handler': <urllib.request.HTTPBasicAuthHandler object at 0x7f466af01f30>, 'context': <ssl.SSLContext object at 0x7f4669d2ba40>}

 fh2.__dict__
{'log': <Logger Fhem (WARNING)>, 'server': 'fhem2', 'port': '8084', 'ssl': True, 'csrf': True, 'csrftoken': '', 'username': 'test1', 'password': 'XYZ', 'loglevel': 1, 'connection': False, 'cafile': '', 'nolog': False, 'bsock': None, 'sock': None, 'https_handler': <urllib.request.HTTPSHandler object at 0x7f466b08ca60>, 'protocol': 'https', 'baseurlauth': 'https://fhem2:8084/', 'baseurltoken': 'https://fhem2:8084/fhem', 'baseurl': 'https://fhem2:8084/fhem?XHR=1&cmd=', 'opener': <urllib.request.OpenerDirector object at 0x7f466b08d270>, 'password_mgr': <urllib.request.HTTPPasswordMgrWithDefaultRealm object at 0x7f466b08e4d0>, 'auth_handler': <urllib.request.HTTPBasicAuthHandler object at 0x7f466b08c670>, 'context': <ssl.SSLContext object at 0x7f4669d2b9c0>}
```

From your code, it seems, the message comes from:
It is starting in line 285 in __init__.py
```
            except URLError as err:
                self.connection = False
                self.log.error(
                    "Failed to send msg, len={}, {}".format(len(buf), err))
                return None
```

**Comment by domschl** [comment]
Blind guesses:
Can you rule out problems with network infrastructure (DNS resolution?), e.g. Do fhem1 and fhem2 resolve to different addresses? Are you using some kind of virtualization?

**Comment by domschl** [comment]
Also please enable logging:

```python
import logging
import fhem

logging.basicConfig(level=logging.DEBUG)
```

**Comment by Killerjoe** [comment]
Hi,
sorry for the delay.
I have added both hosts to /etc/hosts with static ip addresses to rule out DNS resolution issues.
fhem1 is for example 192.168.0.98 and fhem2 is 192.168.0.99
It is currently running on a mini pc system with ubuntu 22.04.02 LTS without Virtualization in use.
I get the same result also on both Raspberry Pi 4, on which fhem1 and fhem2 are running.

I did what you were asking and here is the result:
```
>>> import logging
>>> import fhem
>>> logging.basicConfig(level=logging.DEBUG)
>>> 
>>> fh1 = fhem.Fhem("fhem1", port = "8084", protocol = "https", use_ssl = True, username = "test1", password = 'XYZ', csrf=True)
DEBUG:Fhem:Setting up opener on: https://fhem1:8084/
>>> 
>>> fh2 = fhem.Fhem("fhem2", port = "8084", protocol = "https", use_ssl = True, username = "test1", password = 'XYZ', csrf=True)
DEBUG:Fhem:Setting up opener on: https://fhem2:8084/
>>> fh1.get_readings(name="PCA301.*")
DEBUG:Fhem:Cmd: 
DEBUG:Fhem:Cmd-enc: 
INFO:Fhem:Request: https://fhem1:8084/fhem
ERROR:Fhem:Failed to send msg, len=0, HTTP Error 401: Authorization Required
ERROR:Fhem:No valid answer on send when expecting csrf.
ERROR:Fhem:Failed to get fhem state. Not connected.
{}
>>> fh2.get_readings(name="PCA301.*")
DEBUG:Fhem:Cmd: 
DEBUG:Fhem:Cmd-enc: 
INFO:Fhem:Request: https://fhem2:8084/fhem
DEBUG:Fhem:Sending: jsonlist2 NAME~PCA301.*
DEBUG:Fhem:Cmd: jsonlist2 NAME~PCA301.*
DEBUG:Fhem:Cmd-enc: jsonlist2%20NAME~PCA301.%2A
INFO:Fhem:Request: https://fhem2:8084/fhem?XHR=1&cmd=jsonlist2%20NAME~PCA301.%2A
INFO:Fhem:JSON answer received.
```

**Comment by domschl** [comment]
That will take some time to analyze. I've to first update the project to be able to test the thing:

- the travis CI tests need to be replaced with an existing CI, self-tests
- PY3 only
- modernized logging
- checkout recent FHEM releases (haven't for years)
- python packaging is old standard.

That will take a while. Lots of tech debt did accumulate.

**Comment by domschl** [comment]
If you want to try something:

```python
                self.log.info("Request: {}".format(ccmd))
                if ccmd.lower().startswith("http"):
                    ans = urlopen(ccmd, paramdata, timeout=timeout, context=self.context)
```

`urlopen()` was using the default context, and that might be a global state! The change would be to add `context=self.context` as additional parameter. It will still take me some time to get the whole thing tested... 

Another test would be NOT to use SSL. I guess it then works with two sessions.

**Comment by domschl** [comment]
There's another global state with the user/pwd auth handling! That would explain the 401 error.

**Comment by domschl** [comment]
I've published a new version [`0.7.0` on PyPI](https://pypi.org/project/fhem/0.7.0/) that should fix the problem: There were indeed global states with SSL sessions and especially the authentication handlers that got overwritten when opening multiple connections. I've fixed the self-test stuff with github actions, and added a case for two concurrent connections. It is not exactly your use-case, since that connects twice to the same server, so I would be grateful, if you could test the new package `0.7.0` from PyPI and report if your problem is actually solved?

**Comment by Killerjoe** [comment]
HI domschl,
thank you very much for your efforts, now with version 0.7.0 it is working like expected.
I can connect via ssl to two different instances and get the results back without any issues.
Best Regards,
Jochen
